axelor-production/src/main/java/com/axelor/apps/production/service/batch/BatchFillSopRealValues.java

```diff
@@ -0,0 +1,180 @@
+    this.saleOrderLineRepo = saleOrderLineRepo;
+    this.manufOrderRepo = manufOrderRepo;
+    this.stockLocationLineHistoryRepo = stockLocationLineHistoryRepo;
+    this.periodService = periodService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Override
+  protected void process() {
+    ProductionBatch productionBatch = batch.getProductionBatch();
+    computeSopRealValues(productionBatch);
+  }
+
+  @Transactional
+  protected void computeSopRealValues(ProductionBatch productionBatch) {
+    Company company = productionBatch.getCompany();
+    for (Sop sop : productionBatch.getSopSet()) {
+      computeValuesPerSOP(sop, company);
+    }
+  }
+
+  protected void computeValuesPerSOP(Sop sop, Company company) {
+    ProductCategory productCategory = sop.getProductCategory();
+    for (SopLine sopLine : sop.getSopLineList()) {
+      try {
+        computeRealSalesAndSalesGap(company, productCategory, sopLine);
+        computeRealProduction(company, productCategory, sopLine);
+        computeRealStock(company, productCategory, sopLine);
+        incrementDone();
+      } catch (Exception e) {
+        incrementAnomaly();
+        TraceBackService.trace(e, ExceptionOriginRepository.SOP, batch.getId());
+      }
+    }
+  }
+
+  protected void computeRealSalesAndSalesGap(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealSales(
+        saleOrderLineRepo
+            .all()
+            .filter(
+                "self.saleOrder.statusSelect IN (3,4) AND self.product.productCategory = :productCategory AND self.saleOrder.company = :company AND self.saleOrder.confirmationDateTime BETWEEN :startDate AND :endDate")
+            .bind("productCategory", productCategory)
+            .bind("company", company)
+            .bind("startDate", period.getFromDate())
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(SaleOrderLine::getInTaxTotal)
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
+
+    // compute % sales gap
+    computeSopSalesGap(sopLine);
+  }
+
+  protected void computeRealProduction(
+      Company company, ProductCategory productCategory, SopLine sopLine) {
+    Period period = sopLine.getPeriod();
+    sopLine.setSopRealManuf(
+        manufOrderRepo
+            .all()
+            .filter(
+                "self.statusSelect = :statusSelect AND self.product.productCategory = :productCategory AND self.company = :company AND self.realEndDateT BETWEEN :startDate AND :endDate")
+            .bind("statusSelect", ManufOrderRepository.STATUS_FINISHED)
+            .bind("company", company)
+            .bind("productCategory", productCategory)
+            .bind("startDate", period.getFromDate())
+            .bind("endDate", period.getToDate())
+            .fetchStream()
+            .map(ManufOrder::getQty)
+            .reduce(BigDecimal.ZERO, BigDecimal::add));
```

Comment:
Please, if it is not too difficult, use directly a query (createQuery method) to compute the sum instead of using a fetchStream